pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,360 @@
+    ...    cost = opt.step_and_cost()
+    ...    print(f"Step {step} - cost {cost}")
+    Step 0 - cost -1.3351865007304005
+    Step 1 - cost -1.9937887238935206
+    Step 2 - cost -2.1524234485729834
+    Step 3 - cost -2.1955105378898487
+    Step 4 - cost -2.2137628169764256
+    Step 5 - cost -2.2234364822091575
+    ...
+
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, circuit, stepsize=0.01, restriction=None, exact=False, trottersteps=1):
+
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(f"circuit must be a QNode, received {type(circuit)}")
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        self.hamiltonian = circuit.func().obs
+        if not isinstance(self.hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                f"circuit must return the expectation value of a Hamiltonian,"
+                f"received {type(circuit.func().obs)}"
+            )
+        self.nqubits = len(circuit.device.wires)
+
+        if self.nqubits > 4:
+            warnings.warn(
+                "The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow.",
+                UserWarning,
+            )
+        if restriction is not None and not isinstance(restriction, qml.Hamiltonian):
+            raise TypeError(
+                f"restriction must be a Hamiltonian, received {type(restriction)}"
+            )
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators(restriction)
+        self.exact = exact
+        self.trottersteps = trottersteps
+        self.coeffs, self.observables = self.hamiltonian.terms
+        self.stepsize = stepsize
+
+    def step_and_cost(self):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Returns:
+             float: the objective function output prior to the step.
+
+        """
+        cost = self.circuit()
+        omegas = self.get_omegas()
+        nonzero_idx = np.where(~np.isclose(omegas, 0))[0]
+
+        non_zero_lie_algebra_elements = [self.lie_algebra_basis_names[i] for i in nonzero_idx]
+        non_zero_omegas = [-omegas[i] for i in nonzero_idx]
+
+        lie_gradient = qml.Hamiltonian(
+            non_zero_omegas,
+            [
+                qml.grouping.string_to_pauli_word(ps)
+                for ps in non_zero_lie_algebra_elements
+            ],
+        )
+        new_circuit = append_time_evolution(lie_gradient, self.stepsize, self.trottersteps, self.exact)(
+            self.circuit.func
+        )
+
+        self.circuit = qml.QNode(new_circuit, self.circuit.device)
+        return cost
+
+    def step(self):
+        r"""Update the circuit with one step of the optimizer."""
+        self.step_and_cost()
```

Comment:
Two comments here:

1. It might make sense to swap around the logic of `step` and `step_and_cost` 🤔 

E.g.,

```python
def step_and_cost(self):
    cost = self.circuit()
    self.step()
    return cost
```
This avoids the redundant circuit evaluation for a user just wanting to step.

2. It suddenly comes to mind: say you have performed the optimization, and now want to extract the new circuit. How do you retrieve it?

Some ideas that come to mind:

- `opt.optimized_circuit` is a property with a docstring that returns the optimized QNode, or
- `opt.step` returns the optimized QNode?

Maybe the first is nicer. Either way, it would be good to document here in the `step` docstring that nothing is currently returned, since this is a big deviation from the other optimizers!

Reply:
If I switch around the logic, I run into issues because the circuit needs to be constructed before I can append stuff. I did change the logic so that we return the circuit after step/step_and_cost.